File: app/src/main/java/com/example/waveplayer/fragments/fragment_select_songs/RecyclerViewAdapterSelectSongs.java
package com.example.waveplayer.fragments.fragment_select_songs;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;

import androidx.constraintlayout.widget.ConstraintLayout;

import androidx.recyclerview.widget.RecyclerView;

import com.example.waveplayer.activity_main.ActivityMain;
import com.example.waveplayer.random_playlist.Song;
import com.example.waveplayer.R;

import java.util.List;

public class RecyclerViewAdapterSelectSongs extends RecyclerView.Adapter<RecyclerViewAdapterSelectSongs.ViewHolder> {

    private final FragmentSelectSongs fragmentSelectSongs;

    private List<Song> allSongs;

    public RecyclerViewAdapterSelectSongs(FragmentSelectSongs fragmentSelectSongs) {
        this.fragmentSelectSongs = fragmentSelectSongs;
        ActivityMain activityMain = ((ActivityMain) fragmentSelectSongs.getActivity());
        allSongs = activityMain.getAllSongs();
    }

    public void updateList(List<Song> songs) {
        this.allSongs = songs;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerViewAdapterSelectSongs.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new RecyclerViewAdapterSelectSongs.ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_song, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        final ConstraintLayout linearLayout =
                holder.songView.findViewById(R.id.linear_layout_song_name);
            List<Song> userPickedSongs = fragmentSelectSongs.getUserPickedSongs();
            if (userPickedSongs.contains(allSongs.get(position))) {
                allSongs.get(position).setSelected(true);
                holder.textViewSongName.setBackgroundColor(Color.parseColor("#575757"));
                linearLayout.setBackgroundColor(Color.parseColor("#575757"));
            } else {
                allSongs.get(position).setSelected(false);
                holder.textViewSongName.setBackgroundColor(Color.parseColor("#000000"));
                linearLayout.setBackgroundColor(Color.parseColor("#000000"));
            }
            holder.song = allSongs.get(position);
            holder.textViewSongName.setText(allSongs.get(position).title);
        }

    @Override
    public int getItemCount() {
        ActivityMain activityMain = ((ActivityMain) fragmentSelectSongs.getActivity());
        if(activityMain != null) {
            return allSongs.size();
        }
        return -1;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        public final View songView;
        public final TextView textViewSongName;
        public Song song;

        public ViewHolder(final View view) {
            super(view);
            if (fragmentSelectSongs instanceof FragmentSelectSongs) {
                view.findViewById(R.id.song_handle).setVisibility(View.INVISIBLE);
            }
            songView = view;
            textViewSongName = view.findViewById(R.id.text_view_songs_name);
            final ConstraintLayout constraintLayout = view.findViewById(R.id.linear_layout_song_name);
            final ActivityMain activityMain = ((ActivityMain) fragmentSelectSongs.getActivity());
            if(activityMain != null) {
                view.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (song.isSelected()) {
                            song.setSelected(false);
                            textViewSongName.setBackgroundColor(Color.parseColor("#000000"));
                            constraintLayout.setBackgroundColor(Color.parseColor("#000000"));
                            fragmentSelectSongs.removeUserPickedSong(song);
                        } else {
                            song.setSelected(true);
                            textViewSongName.setBackgroundColor(Color.parseColor("#575757"));
                            constraintLayout.setBackgroundColor(Color.parseColor("#575757"));
                            fragmentSelectSongs.addUserPickedSong(song);
                        }
                    }
                });
            }
        }

    }

}

File: app/src/main/java/io/fourthFinger/pinkyPlayer/fragments/RecyclerViewAdapterSelectSongs.kt
package io.fourthFinger.pinkyPlayer.fragments

import android.annotation.SuppressLint
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import io.fourthFinger.pinkyPlayer.R
import io.fourthFinger.pinkyPlayer.random_playlist.Song

class RecyclerViewAdapterSelectSongs(
    private val listenerCallbackSelectSongs: ListenerCallbackSelectSongs,
    private var allSongs: List<Song>
) : RecyclerView.Adapter<RecyclerViewAdapterSelectSongs.ViewHolder>() {

    interface ListenerCallbackSelectSongs {
        fun songUnselected(song: Song)
        fun songSelected(song: Song)
    }

    @SuppressLint("NotifyDataSetChanged")
    fun updateList(songs: List<Song>) {
        allSongs = songs
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.item_song, parent, false
            )
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val linearLayout: ConstraintLayout = holder.itemView.findViewById(
            R.id.constraint_layout_song_name
        )
        // TODO use color resources
        if (allSongs[position].isSelected()) {
            holder.textViewSongName.setBackgroundColor(Color.parseColor("#575757"))
            linearLayout.setBackgroundColor(Color.parseColor("#575757"))
        } else {
            holder.textViewSongName.setBackgroundColor(Color.parseColor("#000000"))
            linearLayout.setBackgroundColor(Color.parseColor("#000000"))
        }
        holder.song = allSongs[position]
        holder.textViewSongName.text = allSongs[position].title
    }

    override fun getItemCount(): Int {
        return allSongs.size
    }

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val textViewSongName: TextView = view.findViewById(R.id.text_view_songs_name)
        lateinit var song: Song

        init {
            // TODO invisible or gone?
            view.findViewById<View?>(R.id.song_handle).visibility = View.GONE
            val constraintLayout: ConstraintLayout =
                view.findViewById(R.id.constraint_layout_song_name)
            // TODO color resources
            val onClickListener = View.OnClickListener { v: View? ->
                if (song.isSelected()) {
                    song.setSelected(false)
                    // TODO color resources
                    textViewSongName.setBackgroundColor(Color.parseColor("#000000"))
                    constraintLayout.setBackgroundColor(Color.parseColor("#000000"))
                    listenerCallbackSelectSongs.songUnselected(song)
                } else {
                    song.setSelected(true)
                    textViewSongName.setBackgroundColor(Color.parseColor("#575757"))
                    constraintLayout.setBackgroundColor(Color.parseColor("#575757"))
                    listenerCallbackSelectSongs.songSelected(song)
                }
            }
            view.setOnClickListener(onClickListener)
        }
    }

}